query_optimizer/rules/UnnestSubqueries.cpp

```diff
@@ -0,0 +1,720 @@
+        return L::Aggregate::Create(aggregate->input(),
+                                    group_expressions,
+                                    aggregate->aggregate_expressions());
+      }
+      return input_with_no_subqueries;
+    }
+    case L::LogicalType::kFilter: {
+      if (new_input_expressions.empty()) {
+        const L::Filter* filter = static_cast<const L::Filter*>(input_with_no_subqueries.get());
+        return filter->input();
+      }
+      if (has_change) {
+        return input_with_no_subqueries->copyWithNewInputExpressions(new_input_expressions);
+      }
+      return input_with_no_subqueries;
+    }
+    case L::LogicalType::kProject: {
+      if (has_change) {
+        input_with_no_subqueries =
+            input_with_no_subqueries->copyWithNewInputExpressions(new_input_expressions);
+      }
+      if (!inner_attributes->empty()) {
+         const L::Project *project = static_cast<const L::Project*>(input_with_no_subqueries.get());
+         // Insert those inner attributes into the front of the project expression
+         // list rather than the back to help reduce unnecessary selection for
+         // reordering output attributes, since those attributes are usually also
+         // grouping attributes that proceed others in projection.
+         std::vector<E::NamedExpressionPtr> new_project_expressions;
+         const E::UnorderedNamedExpressionSet project_expression_set(project->project_expressions().begin(),
+                                                                     project->project_expressions().end());
+         for (const E::AttributeReferencePtr &inner_attribute : *inner_attributes) {
+           if (project_expression_set.find(inner_attribute) == project_expression_set.end()) {
+             new_project_expressions.emplace_back(inner_attribute);
+           }
+         }
+         new_project_expressions.insert(new_project_expressions.end(),
+                                        project_expression_set.begin(),
+                                        project_expression_set.end());
+         if (new_project_expressions.size() != project->project_expressions().size()) {
+           input_with_no_subqueries = L::Project::Create(project->input(),
+                                                         new_project_expressions);
+         }
+      }
+      return input_with_no_subqueries;
+    }
+    default:
+      if (has_change) {
+        return input_with_no_subqueries->copyWithNewInputExpressions(new_input_expressions);
+      }
+      return input_with_no_subqueries;
+  }
+}
+
+E::ExpressionPtr UnnestSubqueriesForNonRootLogical::eliminateOuterAttributeReference(
+    const E::ExpressionPtr &expression,
+    E::UnorderedAttributeSet *inner_attributes,
+    std::vector<E::AttributeReferencePtr> *probe_join_attributes,
+    std::vector<E::AttributeReferencePtr> *build_join_attributes,
+    std::vector<E::PredicatePtr>* non_hash_join_predicates) {
```

Comment:
Could we be consistent in the pointer sigil style here? Thanks!

`std::vector<E::PredicatePtr> *non_hash_join_predicates`